### PR TITLE
ENH: signal: Add functionality of Complex Chirp waveform

### PR DESCRIPTION
**Pull request by anilgurses**
#### What does this implement/fix?
Added complex option for linear chirp. Now, it's possible to generate complex linear chirps.

### CODE REVIEW

**Review by tupui — CHANGES_REQUESTED** [review]
Thanks for the PR @anilgurses. Can you motivate a bit why this is needed? The PR would also need some tests to ensure the new feature is working as intended.

**Comment by anilgurses** [comment]
Thanks for your feedback @tupui . Complex chirp uses the both complex and real part of the signal(I and Q). If anyone wants to use it on radar or communication systems, they may want to obtain more bandwidth by maintaining the same sample rate. From Nyquist rule, bandwidth is limited to 2*F_s. However, if we use the complex part of the signal as well, bandwidth will be equal to F_s. 

I'll add the missing tests and fix the errors that occurred in the build process.

**Comment by anilgurses** [comment]
Another thing about complex chirp is that it's possible to sweep from negative frequency to positive frequency or otherwise.

File: scipy/signal/_waveforms.py
Comment (tupui) [inline scipy/signal/_waveforms.py:295]: From this description, we should then check that user don't mix `'logarithmic', 'hyperbolic'` with `complex=True` (there should also be a test)

```suggestion
    complex : bool, optional
        Only available if `method` is ``'linear'`` or ``'quadratic'``.
        This parameter makes generated signal complex. When `complex` is True,
        it's possible to use negative frequency as `f0`.
        Default is False.
```

File: scipy/signal/_waveforms.py
Comment (tupui) [inline scipy/signal/_waveforms.py:265]: New params should be keyword only.
```suggestion
def chirp(t, f0, t1, f1, method='linear', phi=0, vertex_zero=True, *,
          complex=False):
```

File: scipy/signal/tests/test_waveforms.py
Comment (tupui) [inline scipy/signal/tests/test_waveforms.py:89]: We would need to test more known values to ensure the function behaves as expected and there are no corner cases it cannot handle.

**Review by tupui — CHANGES_REQUESTED** [review]
I did another pass. For the maths and final decision to add this parameter, we still need someone familiar with this function. @ilayn is this something you can help with?

**Comment by roryyorke** [comment]
It would be good to have a reference to the definition of a complex chirp, as well as give some idea of applications (in electronic and electrical engineering I think at least radar, power systems, and communication).

I'd expect the function to be something like the below, which has the property that the real part of the output when `complex_chirp=True` is equal to the output when `complex_chirp=False`.

I also don't have a reference to support this, though.

```python
def chirp(t, f0, t1, f1, method='linear', phi=0, vertex_zero=True, *, complex_chirp=False):
    phase = _chirp_phase(t, f0, t1, f1, method, vertex_zero)
    phi *= pi / 180
    if complex_chirp:
        return exp(1j * (phase + phi)) # exp already imported in _waveforms.py
    else:
        return cos(phase + phi)
```

There's no need to limit it to just linear and quadratic - you just can't get a sign change in frequency with a logarithmic sweep.  I'm not familiar with hyperbolic sweep, but 

If `chirp` were changed like this, `sweep_poly` should probably be changed too, in much the same way.  I don't think `_chirp_phase` or `_sweep_poly_phase` need to be changed at all.

File: scipy/signal/tests/test_waveforms.py
Comment (anilgurses) [reply scipy/signal/tests/test_waveforms.py:89]: Do you have any suggestions? 

PS. Sorry, this fell through the cracks. If you can guide me through some example scenarios, I can write them.

Comment (anilgurses) [reply scipy/signal/tests/test_waveforms.py:106]: Added one more test. I'll think about what else I can add.

**Comment by anilgurses** [comment]
> It would be good to have a reference to the definition of a complex chirp, as well as give some idea of applications (in electronic and electrical engineering I think at least radar, power systems, and communication).

Good point, the only reference I could find for complex chirp waveform is the [Matlab user manual](https://www.mathworks.com/help/signal/ref/chirp.html). I think we shouldn't limit this specifically to complex chirp. My change is adding complex baseband (sort of) support to the chirp waveform. There is a good post about this at DSP StackExchange [link](https://dsp.stackexchange.com/questions/75526/why-is-complex-baseband-better-than-real-representation). Proakis's digital communications book might be a good reference for this.  

Please let me know what I can do for this PR to be merged. 

PS. Once again, sorry for the delay. I finally had time to look at this.

Thanks,
A.

**Comment by DietBru** [comment]
The STFT  plot of a complex chirp using [scipy.signal.ShortTimeFFT](https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.html#scipy.signal.ShortTimeFFT) can be found  [here](https://scipy.github.io/devdocs/tutorial/signal.html#comparison-with-legacy-implementation). Perhaps it would make sense to put an adapted version into the docstring. I think it is pretty self-explanatory.

Note that creating arbitrary chirps is straightforward, if you remember that the instantaneous frequency is derivative of the phase, e.g., 
```python
n, T = 1000, 1e-3  # number of samples and sampling interval
t = np.arange(n)*T  # time stamps
f_i = 5*t +10  # arbitrary function for instantaneous frequency f_i(t)
phi = np.cumsum(f_i)*T  # numeric integration
z = np.exp(2j*np.pi*phi)  # sinusoid with instantaneous frequency f_i(t)
```
(I did not investigate the given code)

**Comment by anilgurses** [comment]
Thanks for your comment @DietBru. The document you shared generates the chirp waveform from scratch. Therefore, it might not be too relevant to our case here but it's helpful to compare the approach. 

Regarding the documentation, I can add the FFT output of complex chirp, similar to Matlab's example. Does it sound reasonable?

**Comment by DietBru** [comment]
Any chirp will do. :smiley: But why use an FFT if there's a shiny new [STFT](https://scipy.github.io/devdocs/reference/generated/scipy.signal.ShortTimeFFT.html)?
My motivation to write down the derivation in the comment was to hint that the [docstring](https://scipy.github.io/devdocs/reference/generated/scipy.signal.chirp.html) could use some love. I think adding the explicit formulas of the chirp (e.g., $x(t)=\cos(\phi(t)+\phi_0), f_i(t) = \phi'(t)/(2\pi)$, ... ) would add a lot more clarity. Also the Python code could be a bit more expressive, e.g.:
  
 ```python
  phase = _chirp_phase(t, f0, t1, f1, method, vertex_zero) + np.deg2rad(phi)
  return np.exp(1j*phase) if complex else np.cos(phase)
  ```

**Comment by anilgurses** [comment]
Oh, it makes more sense now :sweat_smile:. Let me add those details to the docstring. Thanks for the feedback!

File: scipy/signal/tests/test_waveforms.py
Comment (j-bowhay) [inline scipy/signal/tests/test_waveforms.py:109]: New tests should use `assert_allclose`

File: scipy/signal/tests/test_waveforms.py
Comment (j-bowhay) [inline scipy/signal/tests/test_waveforms.py:95]: same here

File: scipy/signal/tests/test_waveforms.py
Comment (j-bowhay) [inline scipy/signal/tests/test_waveforms.py:90]: new tests should use plain asserts
```suggestion
        assert err < 1e-6
```

**Review by j-bowhay — DISMISSED** [review]
some style nits

File: scipy/signal/_waveforms.py
Comment (j-bowhay) [inline scipy/signal/_waveforms.py:293]: Should we raise an error if this is not the case?

Comment (anilgurses) [reply scipy/signal/_waveforms.py:293]: I'll be updating the docstring and comments. All should be fixed by today

**Comment by anilgurses** [comment]
Well, I realized something strange with the new spectrogram [function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.ShortTimeFFT.spectrogram.html#scipy.signal.ShortTimeFFT.spectrogram). It doesn't support complex numbers. Am I missing something?  @DietBru do you have an idea?


```
TypeError: x must be a real sequence
```

**Comment by anilgurses** [comment]
Thanks for the suggestion! I finished the documentation part and other things. Let me know if something missing.

**Comment by lucascolley** [comment]
@anilgurses it's probably time to put this to the mailing list now

File: scipy/signal/_waveforms.py
Comment (DietBru) [inline scipy/signal/_waveforms.py:408]: Perhaps using linear instead of dB scaling is little more suitable here, since it marginally simplifies the code, i.e.:

```python 
Sx2 = SFT.spectrogram(w)
fig, ax = plt.subplots()
im1 = ax.imshow(Sx2, origin='lower', aspect='auto',
                extent=SFT.extent(len(w)), cmap='plasma')
fig.colorbar(im1, label = "Power Spectral Density")
```
Nitpick: The option `scale_to='psd'` produces a power spectral density not a power as written in the colorbar label.

File: scipy/signal/_waveforms.py
Comment (DietBru) [inline scipy/signal/_waveforms.py:461]: Is there a reason that you do incorporate [this](https://github.com/scipy/scipy/pull/17318#issuecomment-1311561161) or [this](https://github.com/scipy/scipy/pull/17318#issuecomment-1925889886) suggestions to simplify the code? I my opinion it would make the code much more readable.

**Review by DietBru — COMMENTED** [review]
I have two comments:

**Comment by DietBru** [comment]
In the [rendered documentation](https://output.circle-artifacts.com/output/job/c9560a8d-51f2-4a34-a9b0-468bb3683912/artifacts/0/html/reference/generated/scipy.signal.chirp.html), I noticed four unwanted blocks with gray highlighting, which is caused of blocks being indented too far, e.g.:
image
This highlighting appeared due to the updating of the sphinx theme. If you want, you can remove one space from the indentation from those four blocks.

**Comment by anilgurses** [comment]
Is there any update on this? It's almost 2 years since my commit :) If the indentation in the document is the problem, I can easily fix it. I believe indentation helps highlight the description of the point.

**Comment by DietBru** [comment]
The main point is that your addition is hard to understand, as commented [here](https://github.com/scipy/scipy/pull/17318/files#r1491790101).
Concerning the indentation: It is helpful if the documentation layout is consistent across all functions.

**Comment by DietBru** [comment]
Gentle ping to @anilgurses : Do you still have interest in finishing this PR? If not, that is fine as well. I would be happy to take up the remaining work.

**Comment by anilgurses** [comment]
Yes, I am still interested. I was busy with other things. I will try to finish it this week.

**Comment by DietBru** [comment]
I just pushed some changes, which:
* Improved `signal.chirp()` docstr—rendered version can be found [here]
* Simplified signal.chirp() code
* Used `xp_assert_allclose()` in `_waveforms.py`
* Manually rebased onto main to resolve conflicts

I hope it also addresses your change request, @j-bowhay, @tupui 

PS: The test failures seem unrelated.

[here]: https://output.circle-artifacts.com/output/job/66c45c67-76de-4c14-bc7a-76d5d459fedd/artifacts/0/html/reference/generated/scipy.signal.chirp.html

**Review by j-bowhay — COMMENTED** [review]
Docs look really great and changes look fine from a python perspective although the subject area is something I'm not familiar with. Feel free to merge if you're happy @DietBru

**Comment by DietBru** [comment]
Thanks for the feedback, @j-bowhay. On this subject matter is not a lot to know—just that any signal can be represented by real-valued amplitude $a(t)$ and a real-valued phase $ϕ(t)$, i.e., $z(t) = a(t)\  \exp\big(j 2 π ϕ (t)\big)$ (with instantaneous frequency $f_i(t)=dϕ/dt$). A great opportunity to deepen the knowledge is to review PR #20097 :grin: 

Also, thank you @anilgurses for your contribution! I squashed the commits again, to be able to add you as author.
Will merge after the tests complete.

**Comment by anilgurses** [comment]
Thanks for your help @DietBru!

**Pull request merged**
